### PR TITLE
Error controller to handle HTTP errors.

### LINKED ISSUE #310: Redirects with go() don't work in CLI.
The `go()` function is defined in `helpers/http.php` and is used to cleanly switch between controllers with an HTTP redirect. 

However this method does not work in CLI.

Additionally it has the disadvantages that: 
1. the original URL is ~~lost~~ changed and 
2. any error parameters have to be passed as GET parameters.

### PR DESCRIPTION
**Pull request by bugos**
Fixes #310 and #257.

In this pull request:
- I defined a new `ErrorRedirectException` that  can be used instead of `RedirectException` to change the controller withoud reloading the page with an HTTP redirect. 
- I changed the `HTTPErrorException` to extend the new `ErrorRedirextException`. When thrown it changes the controller to the new `HTTPErrorController` as @VitSalis suggested in #257.
- I refactored the `ControllerBase`.

### CODE REVIEW

File: controllers/base.php [hunk 1]
```diff
@@ -57,25 +52,43 @@ protected function getControllerVars( $get, $post, $files, $httpRequestMethod )
...
+                $this->method = $vars[ 'method' ];
+            }
+            try {
+                $idempotence = Form::getRESTMethodIdempotence( $this->method ) === 1 ;
+            }
+            catch ( HTMLFormInvalidException $e ) {
+                throw new HTTPNotFoundException( $this->method . ' is not a valid REST method.' );
+            }
+            if ( $idempotence && $httpRequestMethod != 'POST' ) {
+                    $this->method .= 'View';
```
Comment (dionyziz) [inline controllers/base.php]: nit: indent

File: controllers/base.php [hunk 2]
```diff
@@ -57,25 +52,43 @@ protected function getControllerVars( $get, $post, $files, $httpRequestMethod )
...
-                $_SESSION[ 'user' ] = $user;
             }
         }
-        protected function callWithNamedArgs( $methodReflection, $callable, $vars ) {
+        private function getControllerMethod( $vars, $httpRequestMethod ) {
+            if ( isset( $vars[ 'method' ] ) ) {
+                $this->method = $vars[ 'method' ];
+            }
+            try {
+                $idempotence = Form::getRESTMethodIdempotence( $this->method ) === 1 ;
```
Comment (dionyziz) [inline controllers/base.php]: nit: remove space before `;`

Comment (dionyziz) [inline controllers/base.php]: Don't `===` compare against `1`. `1` is truthy and can be used directly in an `if` condition (as is appropriate in a weakly typed language like PHP). But if you prefer, modify `getRESTMethodIdempotence` to return `true` or `false` instead.

File: controllers/base.php [hunk 3]
```diff
@@ -1,10 +1,11 @@
 <?php
     abstract class ControllerBase {
-        protected $acceptTypes = [];
+        public $acceptTypes = [];
```
Comment (dionyziz) [inline controllers/base.php]: Why `public`? Would the dispatcher ever have need to modify or read these?

File: helpers/http.php [hunk 4]
```diff
@@ -45,7 +45,7 @@ public function __construct( $resourceOrURL = false, $method = false, $args = []
         }
     }
 
-    class HTTPErrorException extends Exception {
+    class HTTPErrorException extends ErrorRedirectException {
```
Comment (dionyziz) [inline helpers/http.php:76]: I'm not sure how semantically an `HTTPErrorException` is-an `ErrorRedirectException`. For example, 404 is not a redirect.

File: index.php [hunk 5]
```diff
@@ -14,15 +14,7 @@
...
-
-        $url = $e->getURL();
-
-        header( 'Location: ' . $config[ 'base' ] . $url );
-    }
-    catch ( HTTPErrorException $e ) {
-        header( $e->header );
-        $e->outputErrorPage();
+    catch ( ErrorRedirectException $e ) {
+        $e->callErrorController(); //should catch exception inside the ControllerBase object.
```
Comment (dionyziz) [inline index.php]: You must also modify the other dispatcher - run.

File: controllers/httperror.php [hunk 6]
```diff
@@ -0,0 +1,11 @@
+<?php
+    class HTTPErrorController extends ControllerBase {
+      public $method = 'create';
```
Comment (dionyziz) [inline controllers/httperror.php]: nit: indentation

File: controllers/httperror.php [hunk 7]
```diff
@@ -0,0 +1,11 @@
+<?php
+    class HTTPErrorController extends ControllerBase {
+      public $method = 'create';
+
+        public function createView( $error, $reason, $header ) {
```
Comment (dionyziz) [inline controllers/httperror.php]: Why are we using the `create` RESTful semantics? I know @VitSalis suggested this, but it doesn't seem that the user is asking for an error to be created here - only viewing an error that has occurred.

File: helpers/http.php (hunk 4 above)
Comment (bugos) [inline helpers/http.php:76]: Would it be better if I throwed an ErrorRedirectException inside the HTTPErrorException constructor?

Comment (dionyziz) [inline helpers/http.php:76]: Why do you need to do that? It's not a redirect.

Comment (bugos) [inline helpers/http.php:76]: The point is that when an `HTTPErrorException` (including `404`) is thrown we want to change(redrect) the controller to the new `HTTPErrorController`. This is done when an ErrorRedirectExceprtion is catched with a `$e->controller` set to `httperror`.

File: run [hunk 8]
```diff
@@ -8,23 +8,26 @@
...
+        }
+        catch ( HTTPErrorException $e ) {
+            die( 'An HTTP error occurred: ' . $e->header . ': ' . $e->reason );
+        }
+        catch ( ErrorRedirectException $e ) {
+            launchController( $e->controller, $e->arguments );
+        }
     }
+
+    exit( launchController( $resource, [ 'method' => $method ], $vars, [], 'POST' ) );
```
Comment (dionyziz) [inline run:32]: Why do you need an `exit` here?

Comment (bugos) [inline run:32]: So that the return value of the `run` script is the return value of the dispatcher. This is how it was before; should we  change it?

Comment (dionyziz) [inline run:32]: No, this is fine, I was confused.

File: helpers/html.php [hunk 9]
```diff
@@ -131,10 +131,10 @@ public static function getRESTMethodIdempotence( $method ) {
                 'update' => 1,
                 'view' => 0
             ];
-            if ( isset( $methods[ $method ] ) ) {
-                return $methods[ $method ];
+            if ( !isset( $methods[ $method ] ) ) {
+                throw new HTMLFormInvalidException( $method );
             }
-            throw new HTMLFormInvalidException( $method );
+            return $methods[ $method ];
```
Comment (dionyziz) [inline helpers/html.php:137]: Nice touch.

File: helpers/http.php [hunk 10]
```diff
@@ -1,4 +1,24 @@
...
+                list( $a, $q ) = explode( ';q=', $a );
+                if ( $q === 0 ) {
+                    continue;
+                }
+            }
+            $acceptTypes[ $a ] = true;
+        }
+        return $acceptTypes;
+    }
+
```
Comment (dionyziz) [inline helpers/http.php:21]: Good refactoring to move this to the separate HTTP helper :+1:

Comment (dionyziz) [inline helpers/http.php:21]: We should unit test this. Do you want to do this in this pull request or merge and work on it separately?

**Pull request merged**